Connect AI with interface

robopc opened the pull request:
This PR connects AI and the Interface more by listening to the AIData that gets published by AI.
This data currently includes robot paths and STP data.

Also, this PR fixes the robot positioning and scaling, which previously did not work/had bugs.

This PR is dependent on both [the PR in Networking](https://github.com/RoboTeamTwente/roboteam_networking/pull/11) and [the PR in Utils](https://github.com/RoboTeamTwente/roboteam_utils/pull/120)

robopc committed:
Connect AI with interface

Now also draw robot paths, and fix robot scaling issue

daw1012345 commented on libs/interface/include/InterfaceControllerClient.h:
```diff
@@ -16,21 +17,22 @@ namespace rtt::Interface {
-        std::unique_ptr<rtt::net::WorldSubscriber> field_subscriber;
+        std::unique_ptr<net::WorldSubscriber> fieldSubscriber;
+        std::unique_ptr<net::AIYellowDataSubscriber> yellowDataSub;
+        std::unique_ptr<net::AIBlueDataSubscriber> blueDataSub;
```
Why all the new subscribers? This could've been combined into a single message (see my comments on the networking repo)

daw1012345 commented on libs/interface/include/InterfaceFieldStateStore.h:
```diff
@@ -8,14 +8,23 @@
     mutable std::mutex mtx;
     proto::State state;
 
+    rtt::AIData yellowAIData;
```
I'm not sure what to think about storing AIData with the field. I imagined the field state store to be only things related to world.

daw1012345 commented on libs/interface/include/InterfaceRobotItem.h:
```diff
@@ -19,9 +23,13 @@ class InterfaceRobotItem: public QGraphicsItem {
+          isYellow(isYellow),
+          storage(storage) {}
+    void triggerUpdate();
+    void updateScale(double fieldWidth, double fieldHeight);
```
Instead of updating and storing the scale everywhere every update cycle, why not send it as part of the update signal?

daw1012345 commented on libs/interface/src/InterfaceFieldRenderer.cpp:
```diff
@@ -48,15 +52,16 @@ void InterfaceFieldRenderer::renderField(QPainter *painter, proto::State info, Q
+    double h_scale =  static_cast<double>(canvasHeight) / fieldHeight;
 
-    return std::min({w_scale, h_scale});
+    this->scale = std::fmin(w_scale, h_scale);
```
why `std::fmin`? std::min is a template, there should be no significant difference if the types of the two parameters are as expected.

daw1012345 committed:
Add missing mutex and disable Qt radio

daw1012345 committed:
Restructure the interface controllers to be more flexible and tailor to the need of the AI and the interface better

* Field data is now parsed lazily (only a string is stored)
* The AI only sends data when it's prompted to
* The interface controller sends data over every second, or when a change is detected (checked every 200 ms)

daw1012345 commented on libs/interface/src/InterfaceFieldStateStore.cpp:
```diff
@@ -15,4 +15,14 @@ proto::State InterfaceFieldStateStore::getState() const {
+    return team == rtt::Team::YELLOW ? this->yellowAIData : this->blueAIData;
+}
 
+void InterfaceFieldStateStore::setAIData(const rtt::AIData &data, rtt::Team team) {
```
This is **not** thread safe. Lock the `mtx` mutex before R/W.

rttwente committed:
Include <optional>

daw1012345 committed:
Merge branch 'feature/iface-ctrl-structure' into feature/connectInterface

# Conflicts:
#	libs/interface/include/InterfaceControllerClient.h
#	libs/interface/include/InterfaceFieldStateStore.h
#	libs/interface/src/InterfaceControllerClient.cpp
#	libs/interface/src/InterfaceFieldScene.cpp
#	libs/interface/src/InterfaceFieldStateStore.cpp
#	libs/interface/src/InterfaceFieldView.cpp

daw1012345 committed:
Merge interface controller refactor into connectInterface

daw1012345 committed:
Code cleanup and optimisations

- Move classes into the rtt::Interface namespace
- Do not parse every proto message that arrives
  - Create helper class for caching proto parsing results
- Do not repaint on every update cycle (repaint only when scale changes)
- Temporarily disable path drawing pending optimisations

daw1012345 committed:
Re-enable robot paths with improved performance

Florisheinen1 reviewed (approved):
Approved by Dawid
